robyn/openapi.py

```diff
@@ -0,0 +1,263 @@
+    components: Components = field(default_factory=Components)
+
+
+@dataclass
+class OpenAPI:
+    """
+    This is the root object of the OpenAPI document.
+
+    @param info: Provides metadata about the API.
+    @param openapi_spec: The content of openapi.json as a dict
+    """
+
+    info: OpenAPIInfo = field(default_factory=OpenAPIInfo)
+    openapi_spec: dict = field(init=False)
+
+    def __post_init__(self):
+        """
+        initializes the openapi_spec dict
+        """
+        self.openapi_spec = {
+            "openapi": "3.0.0",
+            "info": asdict(self.info),
+            "paths": {},
+            "components": asdict(self.info.components),
+            "servers": [asdict(server) for server in self.info.servers],
+            "externalDocs": asdict(self.info.externalDocs) if self.info.externalDocs.url else None,
+        }
+
+    def add_openapi_path_obj(self, route_type: str, endpoint: str, openapi_summary: str, openapi_tags: list, signature: Signature = None):
+        """
+        adds the given path to openapi spec
+
+        :param route_type: the http method as string (get, post ...)
+        :param endpoint: the endpoint to be added
+        :param openapi_summary: short summary of the endpoint (to be fetched from the endpoint defenition by default)
+        :param openapi_tags: tags -- for grouping of endpoints
+        :param signature: the function signature -- to grab the typed dict annotations for query params
+        """
+
+        query_params = None
+
+        if signature and "query_params" in signature.parameters:
+            query_params = signature.parameters["query_params"].default
+
+        modified_endpoint, path_obj = self.get_path_obj(endpoint, openapi_summary, openapi_tags, query_params)
+
+        if modified_endpoint not in self.openapi_spec["paths"]:
+            self.openapi_spec["paths"][modified_endpoint] = {}
+        self.openapi_spec["paths"][modified_endpoint][route_type] = path_obj
+
+    def add_subrouter_paths(self, subrouter_openapi):
+        """
+        adds the subrouter paths to main router's openapi specs
+
+        :param subrouter_openapi: the OpenAPI object of the current subrouter
+        """
+        paths = subrouter_openapi.openapi_spec["paths"]
+
+        for path in paths:
+            self.openapi_spec["paths"][path] = paths[path]
+
+    def get_path_obj(self, endpoint: str, summary: str, tags: list, query_params: TypedDict = None):
+        """
+        get the "path" openapi object according to spec
+
+        :param endpoint: the endpoint to be added
+        :param summary: short summary of the endpoint (to be fetched from the endpoint defenition by default)
+        :param tags: tags -- for grouping of endpoints
+        :param query_params: query params for the function
+
+        :return: the "path" openapi object according to spec
+        """
+        # for converting path param from `:param` to `{param}`
```

Comment:
could you write an example here -

e.g. robyn has paths like `/:url/:etc` whereas openapi requires path like `/{url}/{path}` ?